bin/build.sh

```diff
@@ -5,45 +5,153 @@ set -euo pipefail
+    # architectures or multi-arch images. Instead, they are amd64 only.
+    DOCKER_ARGS=("build" "--platform=linux/amd64")
+    # heroku/heroku:22 and prior images need separate *cnb* variants that
+    # add compatibility for Cloud Native Buildpacks.
+    VARIANTS=("-build:" "-cnb:" "-cnb-build:-build")
+else
+    # heroku/heroku:24 images and beyond are multi-arch (amd64+arm64) images.
+    # Due to weak feature support parity between Docker on Linux and Docker
+    # Desktop building and publishing across platforms has caveats (see the
+    # top of this file).
+    if [ "$containerd_snapshotter" = 0 ] || { [ -n "$PUBLISH_SUFFIX" ] && [ "$docker_container_driver" -eq 0 ]; }; then
+        DOCKER_ARGS=("buildx" "build" "--platform=linux/amd64,linux/arm64")
+    elif [ -z "$PUBLISH_SUFFIX" ] && [ "$docker_container_driver" -ne 0 ]; then
+        DOCKER_ARGS=("buildx" "build" "--platform=linux/amd64")
+        echo "WARNING: heroku-24 and newer images are multi-arch images," \
+            "but this script is building single architecture images" \
+            "due to limitations of the current platform." \
+            "To build a multi-arch image, enable the \`containerd\`" \
+            "snapshotter in Docker Desktop and/or use a \`docker-container\`" \
+            "Docker BuildKit driver."
+    else
+        echo "ERROR: Can't build images with this configuration. Enable" \
+            "the \`containerd\` snapshotter in Docker Desktop, enable" \
+            "the \`docker-container\` driver in Docker, or use this script" \
+            "in build-only mode (don't provide PUBLISH_SUFFIX argument)."
+        exit 1
+    fi
+    # heroku/heroku:24 and beyond images include CNB specific
+    # modifications, so separate *cnb* variants are not created.
+    VARIANTS=("-build:")
+fi
+
+if [ -n "$PUBLISH_SUFFIX" ]; then
+    # If there is a tag suffix, this script is pushing to a remote registry.
+    DOCKER_ARGS+=("--push")
+else
+    # Otherwise, load the image into the local image store.
+    DOCKER_ARGS+=("--load")
+fi
 
 write_package_list() {
     local image_tag="$1"
-    local output_file="${2}/installed-packages.txt"
-    echo '# List of packages present in the final image. Regenerate using bin/build.sh' > "$output_file"
-    docker run --rm "$image_tag" dpkg-query --show --showformat='${Package}\n' >> "$output_file"
+    local dockerfile_dir="$2"
+
+    # Extract the stack version from the dockerfile_dir variable (e.g., heroku-24)
+    local stack_version
+    stack_version=$(echo "$dockerfile_dir" | sed -n 's/^heroku-\([0-9]*\).*$/\1/p')
+
+    local archs=("amd64")
+    # heroku-24 and newer are multiarch. If containerd is available,
+    # the package list for each architecture can be generated.
+    if [ "$stack_version" -ge 24 ]; then
+        if [ "$containerd_snapshotter" = 0 ]; then
+            archs+=(arm64)
+        else
+            echo "WARNING: Generating package list for single architecture." \
```

Comment:
This means we'll have no CI coverage of either the scripts or the packages list for ARM64 on pull requests, until the PR merges to `main` and the publish step runs.

This seems like it will result in us periodically only discovering failures after merge?

Is there any way around this? For example having the non-publishing variant do ` docker build` twice, to different local tags, each with a different `--platform` value?

Reply:
> My current thinking is that we deal with this until we can deploy some actions runners with arm64.

Deferring for now and re-evaluating once the GHA arm54 runners are available sounds good to me :-)

Reply:
(We will just need to be extra conscious of this PR-only CI limitation when authoring/reviewing PRs in the meantime)